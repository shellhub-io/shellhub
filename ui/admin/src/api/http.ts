import { AxiosInstance } from "axios";
import { setupInterceptorsTo } from "./interceptors";
import * as axiosTs from "./client";
import { Configuration } from "./client";
import { BaseAPI } from "./client/base";

const configuration = new Configuration();
configuration.basePath = `${window.location.protocol}//${window.location.host}`;
configuration.accessToken = localStorage.getItem("cloud_token") || "";

// eslint-disable-next-line import/no-mutable-exports
let adminApi = new axiosTs.AdminApi(configuration);
// eslint-disable-next-line import/no-mutable-exports
let cloudApi = new axiosTs.CloudApi(configuration);

export const createNewClient = () => {
  const newConfiguration = new Configuration();
  newConfiguration.basePath = `${window.location.protocol}//${window.location.host}`;
  newConfiguration.accessToken = localStorage.getItem("cloud_token") || "";
  adminApi = new axiosTs.AdminApi(newConfiguration);
  cloudApi = new axiosTs.CloudApi(newConfiguration);
  return { adminApi, cloudApi };
};

declare module "./client/base" {
  interface BaseAPI {
    getAxios(): AxiosInstance;
    getConfiguration(): Configuration | undefined;
    setConfiguration(configuration: Configuration): void;
  }
}

/** Returns the axios instance */
BaseAPI.prototype.getAxios = function getAxios(this: BaseAPI): AxiosInstance {
  return this.axios;
};

/** Returns the configuration */
// eslint-disable-next-line vue/max-len
BaseAPI.prototype.getConfiguration = function getConfiguration(this: BaseAPI): Configuration | undefined {
  return this.configuration;
};

/** Sets the configuration */
// eslint-disable-next-line vue/max-len
BaseAPI.prototype.setConfiguration = function setConfiguration(this: BaseAPI, configuration: Configuration): void {
  this.configuration = configuration;
};

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
setupInterceptorsTo(adminApi.axios);

export { adminApi, cloudApi };
