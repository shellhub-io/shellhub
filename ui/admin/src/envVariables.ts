/* eslint-disable import/prefer-default-export */
export const envVariables = {
  announcementsEnable: process.env.SHELLHUB_ANNOUNCEMENTS === "true",
  tinyMceKey: process.env.TINYMCE_API_KEY,
};
