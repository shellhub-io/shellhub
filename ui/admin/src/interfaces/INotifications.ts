export enum INotificationsSuccess {
  exportNamespaces = "export namespaces",
  exportUsers = "export users",
  addUser = "add user",
  userDelete = "delete user",
  userEdit = "edit user",
  namespaceEdit = "update namespace",
  licenseUpload = "upload license",
  announcementDelete = "delete announcement",
  announcementCreate = "create announcement",
  announcementEdit = "edit announcement",
  configureSaml = "configure SAML"
}

export enum INotificationsError {
  deviceList = "list device",
  deviceDetails = "details device",
  exportUsers = "export users",
  userList = "list users",
  userDetails = "details user",
  userDelete = "delete user",
  userEdit = "edit user",
  errorLoginToken = "token login",
  firewallRuleList = "list firewall",
  firewallRuleDetails = "details firewall",
  sessionList = "list session",
  sessionDetails = "details session",
  dashboard = "dashboard",
  license = "license",
  addUser = "add user",
  namespaceList = "list namespace",
  namespaceDetails = "details namespace",
  namespaceEdit = "update namespace",
  namespaceLoad = "load namespace",
  exportNamespaces = "export namespaces",
  announcementList = "list announcement",
  announcementDelete = "delete announcement",
  announcementCreate = "create announcement",
  announcementEdit = "edit announcement",
  configureSaml = "configure SAML"
}

export enum INotificationsCopy {
  tenantId = "Tenant ID",
  authenticationURL = "Authentication URL",
}
