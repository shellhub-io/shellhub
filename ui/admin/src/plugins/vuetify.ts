// Styles
// import "@mdi/font/css/materialdesignicons.css";
import "vuetify/styles";

// Vuetify
import { createVuetify } from "vuetify";

export default createVuetify({
  theme: {
    defaultTheme: "dark",
  },
});
