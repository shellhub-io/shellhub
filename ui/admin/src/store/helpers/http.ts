import Axios from "axios";
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore //* use while route mapping is not in TypeScript
export default () => {
  const axios = Axios.create({
    baseURL: `${window.location.protocol}//${window.location.host}/admin`,
    headers: {
      Authorization: `Bearer ${localStorage.getItem("cloud_token")}`,
    },
  });

  return axios;
};
