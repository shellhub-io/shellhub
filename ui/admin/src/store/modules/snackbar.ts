import { Module } from "vuex";
import { State } from "./../index";

export interface snackbarMessageAndContentType {
  typeMessage: string;
  typeContent: string;
}

export interface SnackbarState {
  snackbarError: boolean;
  snackbarSuccess: boolean;
  snackbarMessageAndContentType: snackbarMessageAndContentType;
  snackbarCopy: boolean;
}

export const snackbar: Module<SnackbarState, State> = {
  namespaced: true,

  state: {
    snackbarError: false,
    snackbarSuccess: false,
    snackbarMessageAndContentType: { typeMessage: "", typeContent: "" },
    snackbarCopy: false,
  },

  getters: {
    snackbarSuccess: (state) => state.snackbarSuccess,
    snackbarError: (state) => state.snackbarError,
    snackbarMessageAndContentType: (state) => state.snackbarMessageAndContentType,
    snackbarCopy: (state) => state.snackbarCopy,
  },

  mutations: {
    setSnackbarSuccessAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarNoContent: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarSuccess = true;
    },

    setSnackbarSuccessDefault: (state) => {
      state.snackbarMessageAndContentType = { typeMessage: "default", typeContent: "" };
      state.snackbarSuccess = true;
    },

    unsetSnackbarSuccess: (state) => {
      state.snackbarSuccess = false;
    },

    setSnackbarErrorLoadingOrAction: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorCustom: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    setSnackbarErrorDefault: (state) => {
      state.snackbarMessageAndContentType = { typeMessage: "default", typeContent: "" };
      state.snackbarError = true;
    },

    setSnackbarErrorLicense: (state, data) => {
      state.snackbarMessageAndContentType = data;
      state.snackbarError = true;
    },

    unsetSnackbarError: (state) => {
      state.snackbarError = false;
    },

    setSnackbarCopy: (state, value) => {
      state.snackbarMessageAndContentType = { typeMessage: "", typeContent: value };
      state.snackbarCopy = true;
    },

    unsetSnackbarCopy: (state) => {
      state.snackbarCopy = false;
    },
  },

  actions: {
    showSnackbarSuccessAction: ({ commit }, value) => {
      const data = { typeMessage: "action", typeContent: value };
      commit("setSnackbarSuccessAction", data);
    },

    showSnackbarSuccessDefault: ({ commit }) => {
      commit("setSnackbarSuccessDefault");
    },

    unsetShowStatusSnackbarSuccess: ({ commit }) => {
      commit("unsetSnackbarSuccess");
    },

    showSnackbarErrorLoading: ({ commit }, value) => {
      const data = { typeMessage: "loading", typeContent: value };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarNoContent: ({ commit }) => {
      const data = { typeMessage: "no-content", typeContent: "" };
      commit("setSnackbarNoContent", data);
    },

    showSnackbarErrorAction: ({ commit }, value) => {
      const data = { typeMessage: "action", typeContent: value };
      commit("setSnackbarErrorLoadingOrAction", data);
    },

    showSnackbarErrorCustom: ({ commit }, value) => {
      const data = { typeMessage: "custom", typeContent: value };
      commit("setSnackbarErrorCustom", data);
    },

    showSnackbarErrorLicense: ({ commit }, value) => {
      const data = { typeMessage: "licenseRequired", typeContent: value };
      commit("setSnackbarErrorLicense", data);
    },

    showSnackbarErrorDefault: ({ commit }) => {
      commit("setSnackbarErrorDefault");
    },

    unsetShowStatusSnackbarError: ({ commit }) => {
      commit("unsetSnackbarError");
    },

    showSnackbarCopy: ({ commit }, value) => {
      commit("setSnackbarCopy", value);
    },

    unsetShowStatusSnackbarCopy: ({ commit }) => {
      commit("unsetSnackbarCopy");
    },
  },
};
