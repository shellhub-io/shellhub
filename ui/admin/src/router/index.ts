import { createRouter, createWebHistory, NavigationGuardNext, RouteLocationNormalized } from "vue-router";
import Login from "@admin/views/Login.vue";
import Dashboard from "@admin/views/Dashboard.vue";
import Users from "@admin/views/Users.vue";
import SettingsLicense from "@admin/components/Settings/SettingsLicense.vue";
import SettingsAuthentication from "@admin/components/Settings/SettingsAuthentication.vue";
import Namespaces from "@admin/views/Namespaces.vue";
import Settings from "@admin/views/Settings.vue";
import { computed } from "vue";
import useLicenseStore from "@admin/store/modules/license";
import useLayoutStore from "@admin/store/modules/layout";
import useAuthStore from "@admin/store/modules/auth";
import { plugin as snackbar } from "@/plugins/snackbar"; // using direct plugin because inject() doesn't work outside components

const routes = [
  {
    path: "/login",
    name: "login",
    component: Login,
    meta: {
      layout: "SimpleLayout",
      requiresAuth: false,
    },
  },
  {
    path: "/",
    name: "dashboard",
    component: Dashboard,
  },
  {
    path: "/users",
    name: "users",
    component: Users,
  },
  {
    path: "/user/:id",
    name: "userDetails",
    component: () => import("@admin/views/UserDetails.vue"),
  },
  {
    path: "/devices",
    name: "devices",
    component: () => import("@admin/views/Device.vue"),
    redirect: {
      name: "listDevices",
    },
    children: [
      {
        path: "",
        name: "listDevices",
        component: () => import("@admin/views/Device.vue"),
      },
    ],
  },
  {
    path: "/device/:id",
    name: "deviceDetails",
    component: () => import("@admin/views/DeviceDetails.vue"),
  },
  {
    path: "/sessions",
    name: "sessions",
    component: () => import("@admin/views/Sessions.vue"),
  },
  {
    path: "/session/:id",
    name: "sessionDetails",
    component: () => import("@admin/views/SessionDetails.vue"),
  },
  {
    path: "/settings",
    name: "Settings",
    component: Settings,
    redirect: { name: "SettingProfile" },
    children: [
      {
        path: "authentication",
        name: "SettingAuthentication",
        component: SettingsAuthentication,
      },
      {
        path: "license",
        name: "SettingLicense",
        component: SettingsLicense,
      },
    ],
  },
  {
    path: "/firewall-rules",
    name: "firewall-rules",
    component: () => import("@admin/views/FirewallRules.vue"),
  },
  {
    path: "/firewall-rules/:id",
    name: "firewallRulesDetails",
    component: () => import("@admin/views/FirewallRulesDetails.vue"),
  },
  {
    path: "/namespaces",
    name: "namespaces",
    component: Namespaces,
  },
  {
    path: "/namespace/:id",
    name: "namespaceDetails",
    component: () => import("@admin/views/NamespaceDetails.vue"),
  },
  {
    path: "/announcements",
    name: "announcements",
    component: () => import("@admin/views/Announcements.vue"),
  },
  {
    path: "/announcement/:uuid",
    name: "announcementDetails",
    component: () => import("@admin/views/AnnouncementDetails.vue"),
  },
  {
    path: "/new-announcement",
    name: "new-announcement",
    component: () => import("@admin/views/NewAnnouncement.vue"),
  },
];

const router = createRouter({
  history: createWebHistory("/admin/"),
  routes,
});

router.beforeEach(
  async (to: RouteLocationNormalized, from: RouteLocationNormalized, next: NavigationGuardNext) => {
    const licenseStore = useLicenseStore();
    const layoutStore = useLayoutStore();
    const authStore = useAuthStore();

    const requiresAuth = to.meta.requiresAuth ?? true;

    const layout = to.meta.layout || "AppLayout";
    layoutStore.setLayout(layout as string);

    if (!authStore.isLoggedIn && requiresAuth) {
      return next({
        name: "login",
        query: { redirect: to.fullPath },
      });
    }

    if (authStore.isLoggedIn && !to.meta.requiresAuth) {
      const license = computed(() => licenseStore.getLicense);

      try {
        await licenseStore.get();

        if (license.value.expired && to.name !== "SettingLicense") {
          snackbar.showError("Your license has expired. Please update it and try again.");
          return next({ name: "SettingLicense" });
        }
      } catch {
        if (to.name !== "SettingLicense") {
          snackbar.showError("Failed to get your license info. Please check it and try again.");
          return next({ name: "SettingLicense" });
        }
      }
    }

    return next();
  },
);

export default router;
