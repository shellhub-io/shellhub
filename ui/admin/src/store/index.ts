import { InjectionKey } from "vue";
import { createStore, Store, useStore as vuexUseStore } from "vuex";
import { stats, StatsState } from "./modules/stats";

import { UsersState, users } from "./modules/users";
import { AuthState, auth } from "./modules/auth";
import { DevicesState, devices } from "./modules/devices";
import { firewallRules, FirewallRulesState } from "./modules/firewall_rules";
import { namespaces, NamespacesState } from "./modules/namespaces";
import { layout, LayoutState } from "./modules/layout";
import { license, LicenseState } from "./modules/license";
import { publicKeys, PublicKeysState } from "./modules/public_keys";
import { sessions, SessionsState } from "./modules/sessions";
import { snackbar, SnackbarState } from "./modules/snackbar";
import { spinner, SpinnerState } from "./modules/spinner";
import { announcement, AnnouncementState } from "./modules/announcement";
import { instance, InstanceState } from "./modules/instance";

export interface State {
  user: UsersState;
  auth: AuthState;
  devices: DevicesState;
  firewallRules: FirewallRulesState;
  layout: LayoutState;
  license: LicenseState;
  namespaces: NamespacesState;
  publicKeys: PublicKeysState;
  sessions: SessionsState;
  snackbar: SnackbarState;
  spinner: SpinnerState;
  stats: StatsState;
  announcement: AnnouncementState;
  instance: InstanceState;
}

// eslint-disable-next-line symbol-description
export const key: InjectionKey<Store<State>> = Symbol();

export const store = createStore<State>({
  modules: {
    users,
    auth,
    devices,
    firewallRules,
    layout,
    license,
    namespaces,
    publicKeys,
    sessions,
    snackbar,
    spinner,
    stats,
    announcement,
    instance,
  },
});

export function useStore(): Store<State> {
  return vuexUseStore(key);
}
