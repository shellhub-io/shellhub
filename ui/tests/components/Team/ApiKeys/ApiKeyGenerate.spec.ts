import { createVuetify } from "vuetify";
import { DOMWrapper, flushPromises, mount, VueWrapper } from "@vue/test-utils";
import { beforeEach, describe, expect, it, vi } from "vitest";
import MockAdapter from "axios-mock-adapter";
import ApiKeyGenerate from "@/components/Team/ApiKeys/ApiKeyGenerate.vue";
import { namespacesApi, usersApi, apiKeysApi } from "@/api/http";
import { store, key } from "@/store";
import { router } from "@/router";
import { envVariables } from "@/envVariables";
import { SnackbarPlugin } from "@/plugins/snackbar";
import { INotificationsError } from "@/interfaces/INotifications";

const node = document.createElement("div");
node.setAttribute("id", "app");
document.body.appendChild(node);

type ApiKeyGenerateWrapper = VueWrapper<InstanceType<typeof ApiKeyGenerate>>;

describe("Api Key Generate", () => {
  let wrapper: ApiKeyGenerateWrapper;

  const vuetify = createVuetify();

  let mockNamespace: MockAdapter;

  let mockUser: MockAdapter;

  let mockApiKeys: MockAdapter;

  const members = [
    {
      id: "507f1f77bcf86cd799439011",
      username: "test",
      role: "owner",
    },
  ];

  const namespaceData = {
    name: "test",
    owner: "test",
    tenant_id: "fake-tenant",
    members,
    settings: {
      session_record: true,
      connection_announcement: "",
    },
    max_devices: 3,
    devices_count: 3,
    created_at: "",
  };

  const authData = {
    status: "success",
    token: "",
    user: "test",
    name: "test",
    tenant: "fake-tenant",
    email: "test@test.com",
    id: "507f1f77bcf86cd799439011",
    role: "owner",
    mfa: {
      enable: false,
      validate: false,
    },
  };

  const session = true;

  const getKeyResponse = [
    {
      name: "aaaa2",
      tenant_id: "00000-0000-0000-0000-00000000000",
      role: "administrator",
      created_by: "66562f80daba745a106393b5",
      created_at: "2024-06-07T12:10:56.531Z",
      updated_at: "2024-06-07T12:31:03.505Z",
      expires_in: 1720354256,
    },
    {
      name: "aaaa2",
      tenant_id: "00000-0000-0000-0000-00000000000",
      role: "administrator",
      created_by: "66562f80daba745a106393b5",
      created_at: "2024-06-07T12:10:56.531Z",
      updated_at: "2024-06-07T12:31:03.505Z",
      expires_in: 1720354256,
    },
  ];

  beforeEach(async () => {
    const el = document.createElement("div");
    document.body.appendChild(el);
    vi.useFakeTimers();
    localStorage.setItem("tenant", "fake-tenant");
    envVariables.isCloud = true;

    mockNamespace = new MockAdapter(namespacesApi.getAxios());
    mockUser = new MockAdapter(usersApi.getAxios());
    mockApiKeys = new MockAdapter(apiKeysApi.getAxios());

    mockNamespace.onGet("http://localhost:3000/api/namespaces/fake-tenant-data").reply(200, namespaceData);
    mockUser.onGet("http://localhost:3000/api/users/security").reply(200, session);
    mockUser.onGet("http://localhost:3000/api/auth/user").reply(200, authData);
    mockUser.onGet("http://localhost:3000/api/auth/user").reply(200, authData);
    mockApiKeys.onGet("http://localhost:3000/api/namespaces/api-key?page=1&per_page=10").reply(200, getKeyResponse, { "x-total-count": 2 });

    store.commit("auth/authSuccess", authData);
    store.commit("auth/changeData", authData);
    store.commit("namespaces/setNamespace", namespaceData);
    store.commit("security/setSecurity", session);
    store.commit("apiKeys/setKeyList", { data: getKeyResponse, headers: { "x-total-count": 2 } });

    wrapper = mount(ApiKeyGenerate, {
      global: {
        plugins: [[store, key], vuetify, router, SnackbarPlugin],
        config: {
          errorHandler: () => { /* ignore global error handler */ },
        },
      },
    });
  });

  it("Is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("Renders the component", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Data is defined", () => {
    expect(wrapper.vm.$data).toBeDefined();
  });

  it("Renders components", async () => {
    expect(wrapper.find('[data-test="namespace-generate-main-btn"]').exists()).toBe(true);
    await wrapper.findComponent('[data-test="namespace-generate-main-btn"]').trigger("click");
    const dialog = new DOMWrapper(document.body);
    await flushPromises();
    expect(dialog.find('[data-test="namespace-generate-dialog"]').exists()).toBe(true);
    expect(dialog.find('[data-test="namespace-generate-title"]').exists()).toBe(true);
    expect(dialog.find('[data-test="key-name-text"]').exists()).toBe(true);
    expect(dialog.find('[data-test="namespace-generate-date"]').exists()).toBe(true);
    expect(dialog.find('[data-test="successKey-alert"]').exists()).toBe(false);
    expect(dialog.find('[data-test="keyResponse-text"]').exists()).toBe(false);
    expect(dialog.find('[data-test="close-btn"]').exists()).toBe(true);
    expect(dialog.find('[data-test="add-btn"]').exists()).toBe(true);
  });

  it("Successfully Generate Api Key", async () => {
    mockApiKeys.onPost("http://localhost:3000/api/namespaces/api-key").reply(200, { id: "fake-id" });

    const storeSpy = vi.spyOn(store, "dispatch");

    await wrapper.findComponent('[data-test="namespace-generate-main-btn"]').trigger("click");

    await wrapper.findComponent('[data-test="key-name-text"]').setValue("my api key");

    await wrapper.findComponent('[data-test="add-btn"]').trigger("click");
    await flushPromises();
    expect(storeSpy).toHaveBeenCalledWith("apiKeys/generateApiKey", {
      name: "my api key",
      role: "observer",
      expires_at: 30,
      tenant: "fake-tenant",
    });
  });

  it("Fails to Generate Api Key", async () => {
    mockApiKeys.onPost("http://localhost:3000/api/namespaces/api-key").reply(500);

    const storeSpy = vi.spyOn(store, "dispatch");

    await wrapper.findComponent('[data-test="namespace-generate-main-btn"]').trigger("click");
    const dialog = new DOMWrapper(document.body);

    await wrapper.findComponent('[data-test="key-name-text"]').setValue("my api key");

    await wrapper.findComponent('[data-test="add-btn"]').trigger("click");
    await flushPromises();
    expect(storeSpy).toHaveBeenCalledWith(
      "snackbar/showSnackbarErrorAction",
      INotificationsError.generateKey,
    );

    expect(wrapper.vm.errorMessage).toBe("An error occurred while generating your API key. Please try again later.");

    expect(dialog.find('[data-test="failMessage-alert"]').exists()).toBe(true);
  });

  it("Fails to Generate Api Key (400)", async () => {
    mockApiKeys.onPost("http://localhost:3000/api/namespaces/api-key").reply(400);

    await wrapper.findComponent('[data-test="namespace-generate-main-btn"]').trigger("click");
    const dialog = new DOMWrapper(document.body);

    await wrapper.findComponent('[data-test="key-name-text"]').setValue("");

    await wrapper.findComponent('[data-test="add-btn"]').trigger("click");

    await flushPromises();

    expect(wrapper.vm.errorMessage).toBe("Please provide a name for the API key.");

    expect(dialog.find('[data-test="failMessage-alert"]').exists()).toBe(true);
  });

  it("Fails to Generate Api Key (401)", async () => {
    mockApiKeys.onPost("http://localhost:3000/api/namespaces/api-key").reply(401);

    await wrapper.findComponent('[data-test="namespace-generate-main-btn"]').trigger("click");

    const dialog = new DOMWrapper(document.body);

    await wrapper.findComponent('[data-test="key-name-text"]').setValue("my api key");

    await wrapper.findComponent('[data-test="add-btn"]').trigger("click");

    await flushPromises();

    expect(dialog.find('[data-test="failMessage-alert"]').exists()).toBe(true);
  });

  it("Fails to Generate Api Key (409)", async () => {
    mockApiKeys.onPost("http://localhost:3000/api/namespaces/api-key").reply(409);

    await wrapper.findComponent('[data-test="namespace-generate-main-btn"]').trigger("click");
    const dialog = new DOMWrapper(document.body);

    await wrapper.findComponent('[data-test="key-name-text"]').setValue("my api key");

    await wrapper.findComponent('[data-test="add-btn"]').trigger("click");
    await flushPromises();

    expect(wrapper.vm.errorMessage).toBe("An API key with the same name already exists.");

    expect(dialog.find('[data-test="failMessage-alert"]').exists()).toBe(true);
  });
});
