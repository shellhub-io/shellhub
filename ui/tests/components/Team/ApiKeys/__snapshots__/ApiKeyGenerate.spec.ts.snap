// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Api Key Generate > Renders the component 1`] = `
"<div>
  <div aria-describedby="v-tooltip-v-0"><button type="button" class="v-btn v-btn--elevated v-theme--light bg-primary v-btn--density-default v-btn--size-default v-btn--variant-elevated" data-test="namespace-generate-main-btn"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator=""> Generate key </span>
      <!---->
      <!---->
    </button></div>
  <!--teleport start-->
  <!--teleport end-->
  <!---->
  <!---->
</div>"
`;
